pkg/app/piped/executor/custom/custom.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package custom
+
+import (
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/pipe-cd/pipecd/pkg/app/piped/executor"
+	"github.com/pipe-cd/pipecd/pkg/app/piped/toolregistry"
+	"github.com/pipe-cd/pipecd/pkg/config"
+	"github.com/pipe-cd/pipecd/pkg/model"
+)
+
+type deployExecutor struct {
+	executor.Input
+
+	repoDir string
+	appDir  string
+}
+
+type registerer interface {
+	Register(stage model.Stage, f executor.Factory) error
+}
+
+// Register registers this executor factory into a given registerer.
+func Register(r registerer) {
+	f := func(in executor.Input) executor.Executor {
+		return &deployExecutor{
+			Input: in,
+		}
+	}
+	r.Register(model.StageCustomSync, f)
+}
+
+// Execute exec the user-defined scripts in timeout duration.
+func (e *deployExecutor) Execute(sig executor.StopSignal) model.StageStatus {
+	var originalStatus = e.Stage.Status
+	ctx := sig.Context()
+	ds, err := e.TargetDSP.Get(ctx, e.LogPersister)
+	if err != nil {
+		e.LogPersister.Errorf("Failed to prepare target deploy source data (%v)", err)
+		return model.StageStatus_STAGE_FAILURE
+	}
+	e.repoDir = ds.RepoDir
+	e.appDir = ds.AppDir
+	timeout := e.StageConfig.CustomSyncOptions.Timeout.Duration()
+
+	c := make(chan model.StageStatus, 1)
+	go func() {
+		c <- e.executeCommand(e.StageConfig.CustomSyncOptions)
+	}()
+
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+
+	for {
+		select {
+		case result := <-c:
+			return result
+		case <-timer.C:
+			e.LogPersister.Errorf("Canceled because of timeout")
+			return model.StageStatus_STAGE_FAILURE
+
+		case s := <-sig.Ch():
+			switch s {
+			case executor.StopSignalCancel:
+				return model.StageStatus_STAGE_CANCELLED
+			case executor.StopSignalTerminate:
+				return originalStatus
+			default:
+				return model.StageStatus_STAGE_FAILURE
+			}
+		}
+	}
+}
+
+func (e *deployExecutor) executeCommand(opts *config.CustomSyncOptions) model.StageStatus {
+
+	binDir := toolregistry.DefaultRegistry().GetBinDir()
+	pathFromOS := os.Getenv("PATH")
+
+	path := binDir + ":" + pathFromOS
+	envs := make([]string, 0, len(opts.Envs))
+	for key, value := range opts.Envs {
+		envs = append(envs, key+"="+value)
+	}
+	cmd := exec.Command("/bin/sh", "-c", opts.Run)
+	e.LogPersister.Infof("Runnnig commands...")
+	for _, v := range strings.Split(opts.Run, "\n") {
+		if v != "" {
+			e.LogPersister.Infof("   %s (env: %v)", v, envs)
```

Comment:
Do we need to print `(env: %v)` here? I mean, we may in case envs is empty and always print this doesn't make sense.

Reply:
I think to show env is bad for visibility, too. And in case users want to see environment variables, users can use echo command to see it.

Reply:
Then lets remove it
```suggestion
			e.LogPersister.Infof("   %s", v)
```